### DISCUSSION TITLE
Text tool

### DISCUSSION THREAD

**Discussion by OverloadedOrama** [Orama-Interactive/Pixelorama#589] [Ideas]
**Feature/enhancement description:**
A text tool that lets the user type text into the canvas, and make it part of the image. You can track the progress in the [`text-tool`](https://github.com/Orama-Interactive/Pixelorama/tree/text-tool) branch. Keep in mind that it's still a WIP and will not work properly yet. Planned features:

- [x] Turn font data into pixels in the current cel image itself. 
- [x] Change font. Pick from Roboto and DroidSansFallback, which already exist in Pixelorama, and from system fonts.
- [x] Change text size.
- [x] Save tool options, and check if the loaded fonts still exist. If they don't, remove them from being options.
- [x] Undo/redo.
- [ ] Ability to load fonts as ttf/otf/woff files.
- [ ] Implement non-destructive text editing when vector layers are added.

**Comment by Variable-ind** [comment]
About the `Ability to load fonts from the Web version.`, maybe it could be set to open a simple dialogue saying `ready to pick dropped fonts` (with some funny picture perhaps) instead of file dialog for web version

**Comment by Variable-ind** [comment]
Also i made a variation of the text tool <[Here](https://github.com/Variable-ind/Pixelorama/tree/new-text-tool)>. What do you think about it?

https://user-images.githubusercontent.com/77773850/142255796-b8b975c5-2b1a-4f51-9713-f01ca53f6cab.mp4

**Comment by ManicMarrc** [comment]
Is there any news on this or is it already been implemented but I don't know it or is it discontinued?

**Reply by Variable-ind** [reply]
Hy at the moment it is available in 3d layers, but another implementation is also being worked on

**Reply by piotrtrkostrzewski94** [reply]
Hello,
Is it likely that the text tool will be available for 2D in the near future?
I don't know if it's helpful, but since Godot has the option to add text in the Inspector for the Button node, maybe the code for this Godot function (after any changes) will work for Pixelorama.

**Comment by OverloadedOrama** [comment]
The current plan for the text tool is to include it with vector layers. Vector layers will contain data for shapes and text instead of pixels, which will allow them to be non-destructive, meaning users will be able to change them whenever they wish, and do stuff like resizing a shape, changing the text whenever they want, etc. My idea for this is to store the vector data in SVG form, and convert it to image using Godot's [Image.load_svg_from_string()](https://docs.godotengine.org/en/stable/classes/class_image.html#class-image-method-load-svg-from-string) method. Godot is using [ThorVG](https://github.com/thorvg/thorvg) to handle SVG rendering. However, there are some limitations, such as [lack of aliased shape rendering](https://github.com/thorvg/thorvg/issues/1869), and [no SVG text support](https://github.com/thorvg/thorvg/issues/1897). Once these are implemented, we could create vector layers and finally implement the text tool (although we also need to create UI and gizmos for all of that, like I did for 3D layers).

If ThorVG does not implement these features, or their implementation is not good enough for our needs, we'd have to resort on creating our own way to rasterize vector shapes and text, which can probably be done with Viewports, but I feel like letting Godot/ThorVG handle rasterization is the better solution.

**Reply by piotrtrkostrzewski94** [reply]
I understand,thank you for your answer.
And the option to do it simply as pxo with the option to export to SVG?  If someone would like to create a button (e.g. me) for game in Godot in Pixelorama, all they need to do is export it to png.

**Reply by OverloadedOrama** [reply]
Exporting vector layers as SVG will be possible, if that's what you are asking. It's one of the reasons I am thinking of storing the vector data in SVG form, to allow easy exporting to SVG is users want to.

**Reply by piotrtrkostrzewski94** [reply]
I expressed myself wrong.  I meant that in Godot there is a Button, we can add text in the Inspector, but the button doesn't look very nice.  It's better to use TextureButton but it doesn't have the ability to add text without a Label.  I thought it would be possible to create a text functionality for Pixelorama in the style of Godot in the Inspector.  For Godot, PNG is enough but SVG can also be useful.
![29151458-7dffb1be-7dab-11e7-832a-c8a787ed0cbd](https://github.com/Orama-Interactive/Pixelorama/assets/149613701/e1d80d22-e58f-4d79-9bdb-5b5236d9fedc)

**Reply by Variable-ind** [reply]
Hy, do you mean something that is specific to the button?(I mean Godot also has better ways to add text, like Label2d).
by adding text in inspector, do you mean something like what I showed in a video (few comments up) where i typed the text in properties panel.?

https://github.com/Orama-Interactive/Pixelorama/assets/77773850/4f00b227-a446-456f-9749-a2b963ec6b1c

**Reply by piotrtrkostrzewski94** [reply]
Thank you very much for suggestion.
I think this solution will suit me. 
I will check it in combination with the use of a pencil.  
Has it made it to the stable version?

**Reply by Variable-ind** [reply]
not yet, but if that's all you want (a way to write  text in pixelorama and export as PNG) then it should also be possible in 3d layer, which is available in stable version (it can be switched to perspective mode)
For example see the video below :grin:

https://github.com/Orama-Interactive/Pixelorama/assets/77773850/16e2d424-1178-4a0f-8ce3-e796c5813079

**Reply by piotrtrkostrzewski94** [reply]
Thank you very much 🥇 
Text with "visual layer" :)
